electrum: 2.9.3 -> 3.0.2

Moredread opened the pull request:
###### Motivation for this change

Upgrading electrum to the latest version. As it doesn't support Python 2 anymore, it is now built against Python 3. For this, _jsonrpclib-pelix_ is needed, a fork of _jsonrpclib_, that works with Python 3. See the first of the commit in the PR. If this is acceptable, would you, @joachifm, also be the maintainer of the new package?

I couldn't test it built against master, as I get an error message when I run the binary. I think it's an incompatibility of the QT5 runtime library. See the comment below for details. I backported it to 17.09 (only needs a later version of protobuf, e.g. protobuf3_2), and there it works fine.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Moredread commented:
I couldn't test it built against master, as I get an error message when I run the binary on a 17.09 system:

```
> QT_DEBUG_PLUGINS=1 ./result/bin/electrum

(... skipped ...)

Got keys from plugin meta data ("xcb")
QFactoryLoader::QFactoryLoader() checking directory path "/nix/var/nix/profiles/default/lib/qt-5.9/plugins/platforms" ...
QFactoryLoader::QFactoryLoader() checking directory path "/run/current-system/sw/lib/qt-5.9/plugins/platforms" ...
QFactoryLoader::QFactoryLoader() looking at "/run/current-system/sw/lib/qt-5.9/plugins/platforms/libqminimal.so"
Got keys from plugin meta data ("minimal")
QFactoryLoader::QFactoryLoader() looking at "/run/current-system/sw/lib/qt-5.9/plugins/platforms/libqoffscreen.so"
Got keys from plugin meta data ("offscreen")
QFactoryLoader::QFactoryLoader() looking at "/run/current-system/sw/lib/qt-5.9/plugins/platforms/libqvnc.so"
Got keys from plugin meta data ("vnc")
QFactoryLoader::QFactoryLoader() looking at "/run/current-system/sw/lib/qt-5.9/plugins/platforms/libqxcb.so"
Got keys from plugin meta data ("xcb")
QFactoryLoader::QFactoryLoader() checking directory path "/nix/store/166s7l3yjqfc8dj5hfqjb09dbfvp1850-python3-3.6.3/bin/platforms" ...
Cannot load library /nix/store/i1mgik5x2f8akjwhjcd0r62259s89592-qtbase-5.9.1-bin/lib/qt-5.9/plugins/platforms/libqxcb.so: (/nix/store/h1a1ncbkkhapzm0509plqjlfrgxw22f3-glibc-2.25-49/lib/libresolv.so.2: symbol __res_maybe_init, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference)
QLibraryPrivate::loadPlugin failed on "/nix/store/i1mgik5x2f8akjwhjcd0r62259s89592-qtbase-5.9.1-bin/lib/qt-5.9/plugins/platforms/libqxcb.so" : "Cannot load library /nix/store/i1mgik5x2f8akjwhjcd0r62259s89592-qtbase-5.9.1-bin/lib/qt-5.9/plugins/platforms/libqxcb.so: (/nix/store/h1a1ncbkkhapzm0509plqjlfrgxw22f3-glibc-2.25-49/lib/libresolv.so.2: symbol __res_maybe_init, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference)"
This application failed to start because it could not find or load the Qt platform plugin "xcb"
in "".

Available platform plugins are: minimal, offscreen, vnc, xcb.

Reinstalling the application may fix this problem.
[1]    22886 abort      QT_DEBUG_PLUGINS=1 ./result/bin/electrum
```

FRidh commented on pkgs/applications/misc/electrum/default.nix:
```diff
@@ -35,21 +35,23 @@ python2Packages.buildPythonApplication rec {
     # $out/nix ...
-    mv $out/lib/python2.7/site-packages/nix/store"/"*/share $out
-    rm -rf $out/lib/python2.7/site-packages/nix
+    mv $out/lib/python3.6/site-packages/nix/store"/"*/share $out
```
${python.sitePackages}

Moredread replied:
I've changed it

FRidh commented on pkgs/development/python-modules/jsonrpclib-pelix/default.nix:
```diff
@@ -0,0 +1,22 @@
+    description = "JSON RPC client library - Pelix compatible fork";
+    homepage = https://pypi.python.org/pypi/jsonrpclib-pelix/;
+    license = lib.licenses.asl20;
+    maintainers = [ lib.maintainers.joachifm ];
```
Is @joachifm maintaining this expression?

Moredread replied:
They haven't replied yet. What's the procedure for finding a maintainer for necessary new dependencies?

disassembler replied:
You can make yourself the maintainer or leave it empty (preference on the former).

Moredread replied:
I've removed the maintainer line for now, I'm not sure I can provide the level of attention needed atm.

joachifm replied:
I only care to the extent that it affects electrum, so wouldn't be maintaining it per se. It used to be internal to the electrum expression for that reason.

Moredread replied:
I've added myself as maintainer afterall.

Moredread committed:
maintainers: Add moredread

Moredread committed:
pythonPackages.jsonrpclib-pelix: init at 0.3.1

This fork of jsonrpclib supports Python 3 and is necessary for electrum
from version 3.0.0 onwards.

Adding myself - moredread - as maintainer.

Moredread committed:
electrum: 2.9.3 -> 3.0.2

Moredread commented:
Have added myself as maintainer, and updated the package to 3.0.2, which was released in the meantime.

nonsequitur commented:
For those who want to run Electrum 3 now, here's a quick derivation with pinned dependencies:
https://gist.github.com/nonsequitur/1782996516bdf39b16e647842b1ae4cf
(The proposed patch still doesn't run on master: `xcb` not found)

Moredread commented:
@nonsequitur do you mean my PR or your gist that isn't working with master? Are you running NixOS 17.09 or master on your system?

nonsequitur commented:
@Moredread I'm running NixOS revision [eafd703](https://github.com/NixOS/nixpkgs/tree/eafd703a63a598a33028aca6f1349544dfd09b65) which is in the 17.09 channel.
The electrum build from my gist is working on that system.
When changing the nixpkgs version in the gist to `master`, I'm getting the `xdg` error.
I haven't tried running your PR (or the `master` gist) with NixOS `master` installed, but that shouldn't make a difference.
(I've clarified my original post.)

joachifm commented:
Any way to move this ahead?

Moredread commented:
I think the only issue left is whether it really doesn't work on unstable or if there is an issue when using 17.09 as the base system. I.e. someone needs to find out why the xcb plugin loading doesn't work when build against master.

joachifm commented:
Similar libc errors reported in https://github.com/NixOS/nixpkgs/issues/32273 and https://github.com/NixOS/nixpkgs/issues/32065

joachifm commented:
Basically, the problem is a runtime impurity, which can be "fixed" on non-NixOS by setting `QT_PLUGIN_PATH` and `LD_LIBRARY_PATH` appropriately. I've merged this in the interim. Thank you

joachifm commented:
(The above worked for me, anyways)

Moredread commented:
@joachifm Thanks :)

I think it might be a good idea to backport it to the 17.09 channel, as pre-3.0 has problems with SegWit addresses? As the issue with QT/xcb didn't happen on the older channel, that shouldn't be a blocker.

joachifm commented:
If there are problems with the version on stable, a bump makes sense to me.

joachifm commented:
Hm, OTOH I see that SegWit support was *added* in 3.0, so under a strict interpretation that'd count as a new feature not a bug fix, or is there a specific issue with the 2.9 series? (I'm not really clued into SegWit or electrum's support for it).

Moredread commented:
@joachifm I had crashes with watching wallets, that had SegWit transactions, but I have to check whether it was the latestst 2.X release

joachifm commented:
@Moredread I see. If you can ascertain whether there are issues with 2.9, please open a cherry-pick PR if you have the time, or let me know and I'll pick it when I can.

joachifm commented:
Meanwhile, electrum 3.x has been backported to fix a security issue.
